### ISSUE TITLE
LWRP non-string default values of new_resource are missing after hash conversion in chef client>12.4.3

### ISSUE THREAD

**Issue by wp-vtuska** [chef/chef#4127]
The following code doesn't work in chef client 12.5.1 https://gist.github.com/wp-vtuska/9b093542a6e03f37e709
It is only a gutted example. We are using the hash format of the new_resource object as a template variable. It worked before 12.5.1. 

It looks like the undefined non-string default resource attribute values are missing from instance_variables even if we tried to access them explicitly before the object hash conversion. Is it an intentional modification in the newer chef client?

Current workarounds:
1. new_resource.variable = new_resource.variable
2. using only string attribute values

**Comment by thommay** [comment]
@jkeiser

**Comment by coderanger** [comment]
I would vote this to be note a bug. The output from `to_hash` is not specified anywhere other than that it should be compatible with the rest of the `to_json` and `json_create` suite, which it is. We could certainly improve it as a feature change, but we should probably agree on the use cases for those methods first.

**Comment by lamont-granquist** [comment]
This looks like a side effect of not initializing the default when the property is declared but having it be more lazily evaluated in the Chef::Property#get method.  I'm kind of inclined towards this being a desirable way to have properties evaluated and that this is not a bug.

**Comment by coderanger** [comment]
Yeah, if this was going to be reworked, that whole suite of functions would have to grow much more explicit awareness of the new properties class. Not against that, but its a non-trivial addition.

**Comment by jkeiser** [comment]
It would be reasonable, IMO, to have a `.to_hash` that includes defaulted values, for people who want to do that. `.to_hash(include_defaults: true)`

Not too hard, either.

**Comment by coderanger** [comment]
Very :-1: on having that be configurable, we should pick semantics and run with them, and that should be driven on what that method is supposed to be used for :)

**Comment by jkeiser** [comment]
All I'm sayin' is it's sometimes useful to get a hash that contains all the values, including defaults; even though most of the time you want just the ones the user really set.

**Comment by coderanger** [comment]
Can you illustrate those times? I've purposefully avoided using `to_hash` as I always read it as being an internal format that only matters wrt `json_create`. If it is going to have formal semantics we need to lay them out formally.

**Comment by jkeiser** [comment]
"I want to print out all the values that will be applied when creating this resource."

**Comment by jkeiser** [comment]
(Totally cool with this not being `.to_hash`.)

**Comment by lamont-granquist** [comment]
I think I'm :-1: on it being a flag to `.to_hash`

I think if we want to fix it that .to_hash should just call a method which walks the list of properties and vivifies them before rendering the object.

**Comment by coderanger** [comment]
It doesn't need to vivify them specifically, but it should get the value through the property system rather than inspecting instance variables.

**Comment by stevendanna** [comment]
I am not sure it has to be `to_hash` but I definitely think we should be able to export the state of a resource including all of the defaults into a data structure that can easily be passed to other code/systems.

**Comment by lamont-granquist** [comment]
^ based on what I know about what people have been talking about internally, i believe we're gonna wind up hitting this ourselves eventually anyway...

**Comment by jkeiser** [comment]
I agree; but I think when we hit it, a simple hash isn't going to cut it. If you export this:

```
file '/x.txt' do
  content 'hi'
end
```

As this:

```
{
  'name' => '/x.txt',
  'content' => 'hi',
  'mode' => '0644'
}
```

You are losing super critical information (because the user did not in fact want the mode to be modified to 0644). Exporting a resource with its defaults only helps you predict the values you want from a *new* resource, not an existing one. We'd need to export as a structure that tells you which values the user set, and which values are defaults.

**Comment by jkeiser** [comment]
@wp-vtuska if you put this method on your resource it will do what you want:

```ruby
def to_hash_with_defaults
  result = {}
  self.class.properties.each_key do |name|
    result[name] = send(name)
  end
  result
end
```

**Comment by lamont-granquist** [comment]
Ah, yeah, default-vs-desired-vs-current state rears its ugly head yet again...

**Comment by wp-vtuska** [comment]
@jkeiser thanks a lot!

**Comment by wp-vtuska** [comment]
@jkeiser It works in the new chef client(12.5.1) but it's not compatible with the older one(12.4.3). Am I wrong? Am I supposed to make my cookbooks/recipes smarter and check the runtime environment...?
Is this the recommended way?

**Comment by lamont-granquist** [comment]
This probably should be considered a regression.

Better example code of the use case is in #4174:

https://github.com/chasebolt/chef-template-bug/blob/master/cookbooks/buggy_template/providers/default.rb#L9-L12

Its a little bit lowut, but it did work before 12.5.1

**Comment by lamont-granquist** [comment]
Looks like @chasebolt deleted his example cookbook.  The idea is to have a template resource that passes in the properties off the new_resource:

```ruby
template "/whatever" do
  source "whatever.erb"
  parameters new_resource.to_hash
end
```

Which is a little bit lolwut, but is convenient shorthand which we broke.

**Comment by chasebolt** [comment]
I went ahead and added the example cookbook back.

https://github.com/chasebolt/chef-template-bug

**Comment by jkeiser** [comment]
Backcompat = king ... I prefer the new behavior, but I'll get this fixed.

**Comment by jkeiser** [comment]
Fix is in #4227.
